// react-multi-select-component/dist/cjs/select-panel/select-list.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _useMultiSelect = require("../hooks/use-multi-select");

var _selectItem = _interopRequireDefault(require("./select-item"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This component represents an unadorned list of SelectItem (s).
 */
const SelectList = ({
  options,
  onClick,
  skipIndex
}) => {
  const {
    disabled,
    value,
    onChange,
    ItemRenderer
  } = (0, _useMultiSelect.useMultiSelect)();

  const handleSelectionChanged = (option, checked) => {
    if (disabled) return;
    onChange(checked ? [...value, option] : value.filter(o => o.value !== option.value));
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, options.map((o, i) => {
    const tabIndex = i + skipIndex;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: (o == null ? void 0 : o.key) || i
    }, /*#__PURE__*/_react.default.createElement(_selectItem.default, {
      tabIndex: tabIndex,
      option: o,
      onSelectionChanged: c => handleSelectionChanged(o, c),
      checked: !!value.find(s => s.value === o.value),
      onClick: e => onClick(e, tabIndex),
      itemRenderer: ItemRenderer,
      disabled: o.disabled || disabled
    }));
  }));
};

var _default = SelectList;
exports.default = _default;
